### PR TITLE
[FLINK-21100][coordination] Add DeclarativeScheduler 

### PR DESCRIPTION
**Pull request by zentol**
Based on #14909 and #14910.

Implements the `DeclarativeScheduler`. It primarily connects the various `State` classes, the `SlotAllocator` and `ScaleUpController` to create a new streaming-only scheduler that can automatically scale down jobs if not enough slots are available, and scale up automatically once these slots are available.

Contains various other smaller change; one of note is a change in the lifecycle of the declarative `SharedSlot` implementation where the external release callback is now only called during normal execution, i.e., if all logical slots were properly returned. The goal was to make the code easier to reason about, and prevent multiple `release` calls.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StateWithExecutionGraph.java [hunk 1]
```diff
@@ -106,14 +107,17 @@
                                 context.getMainThreadExecutor()));
     }
 
+    @VisibleForTesting
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StateWithExecutionGraph.java:110]: Why did you add this annotation? Isn't this method also used by the states itself

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/SchedulerNGFactoryFactory.java [hunk 2]
```diff
@@ -19,29 +19,32 @@
...
 
-    public static final String SCHEDULER_TYPE_NG = "ng";
-
     private SchedulerNGFactoryFactory() {}
 
     public static SchedulerNGFactory createSchedulerNGFactory(final Configuration configuration) {
-        final String schedulerName = configuration.getString(JobManagerOptions.SCHEDULER);
-        switch (schedulerName) {
-            case SCHEDULER_TYPE_NG:
+        final JobManagerOptions.SchedulerType schedulerType =
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/SchedulerNGFactoryFactory.java:35]: Why are you not checking the JobType, and fall back to the DefaultScheduler if it's batch?

Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/SchedulerNGFactoryFactory.java:35]: See: https://github.com/tillrohrmann/flink/commit/f28532f696d39db88fbcf44006fe40e3316cafd1#diff-e150cbd5bd3e3618a9a6e8996f7b556772a80efe96d5571f6b72dfb2d20cce1cR46

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java [hunk 3]
```diff
@@ -0,0 +1,882 @@
...
+            JobMasterPartitionTracker partitionTracker,
+            RestartBackoffTimeStrategy restartBackoffTimeStrategy,
+            ExecutionDeploymentTracker executionDeploymentTracker,
+            long initializationTimestamp,
+            ComponentMainThreadExecutor mainThreadExecutor,
+            FatalErrorHandler fatalErrorHandler,
+            JobStatusListener jobStatusListener)
+            throws JobExecutionException {
+
+        this.jobInformation = new JobGraphJobInformation(jobGraph);
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java:187]: at least while we are running more and more tests on the decl. scheduler, I wonder if it makes sense to add an check if the jobGraph contains no blocking exchanges: https://github.com/tillrohrmann/flink/commit/f28532f696d39db88fbcf44006fe40e3316cafd1#diff-a1370001c63af7543847c55c658157fbff34721318bb20cd0509129decf5a22eR184

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java [hunk 4]
```diff
@@ -0,0 +1,882 @@
...
+            LOG.debug(
+                    "Transition from state {} to {}.",
+                    state.getClass().getSimpleName(),
+                    newState.getClass().getSimpleName());
+
+            State oldState = state;
+            oldState.onLeave(newState.getClass());
+
+            state = newState;
+            newState.onEnter();
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java:874]: Will we consider removing onEnter() separately?

**Review by rmetzger — COMMENTED** [review]
I made a first pass over the change

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java [hunk 5]
```diff
@@ -0,0 +1,882 @@
...
+    public void runIfState(State expectedState, Runnable action, Duration delay) {
+        componentMainThreadExecutor.schedule(
+                () -> runIfState(expectedState, action), delay.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    // ----------------------------------------------------------------
+
+    private void transitionToState(State newState) {
+        if (state != newState) {
+            LOG.debug(
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java:865]: I'd say debug is fine because users won't be able to make sense of scheduler states. In fact I think it could even be confusing if the job fails but the schedulers transitions to "fnished" (opposed to failed).

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/SchedulerNGFactoryFactory.java (hunk 2 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/SchedulerNGFactoryFactory.java:35]: mistake during the rebase

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java (hunk 3 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java:187]: It could be that I missed that entire commit :/

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java (hunk 4 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java:874]: yes. As part of that I'd also like to enforce that we cannot transitions into the same state again, which should alleviate the concern about changed semantics.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StateWithExecutionGraph.java (hunk 1 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StateWithExecutionGraph.java:110]: Since the scheduler should not be able to access them, I assumed this was done for testing purposes.
If they were simply made accessible to sub-classes, then they should be made protected to convey this.

**Comment by zentol** [comment]
The DeclarativeSchedulerClusterITCase can currently fail because runIfState _sometimes_ does not run in the main thread.

```
16490 [pool-4-thread-1] WARN  org.apache.flink.runtime.rpc.MainThreadValidatorUtil [] - Violation of main thread constraint detected: expected <Thread[main,5,main]> but running in <Thread[pool-4-thread-1,5,main]>.
java.lang.Exception: Violation of main thread constraint detected: expected <Thread[main,5,main]> but running in <Thread[pool-4-thread-1,5,main]>.
	at org.apache.flink.runtime.rpc.MainThreadValidatorUtil.isRunningInExpectedThread(MainThreadValidatorUtil.java:73) ~[classes/:?]
	at org.apache.flink.runtime.concurrent.ComponentMainThreadExecutorServiceAdapter.lambda$new$0(ComponentMainThreadExecutorServiceAdapter.java:54) ~[test-classes/:?]
	at org.apache.flink.runtime.concurrent.ComponentMainThreadExecutorServiceAdapter.assertRunningInMainThread(ComponentMainThreadExecutorServiceAdapter.java:84) ~[test-classes/:?]
	at org.apache.flink.runtime.executiongraph.ExecutionGraph.assertRunningInJobMasterMainThread(ExecutionGraph.java:1469) ~[classes/:?]
	at org.apache.flink.runtime.executiongraph.ExecutionGraph.transitionState(ExecutionGraph.java:1052) ~[classes/:?]
	at org.apache.flink.runtime.executiongraph.ExecutionGraph.transitionState(ExecutionGraph.java:1044) ~[classes/:?]
	at org.apache.flink.runtime.executiongraph.ExecutionGraph.transitionToRunning(ExecutionGraph.java:834) ~[classes/:?]
	at org.apache.flink.runtime.scheduler.declarative.DeclarativeScheduler.createExecutionGraphWithAvailableResources(DeclarativeScheduler.java:580) ~[classes/:?]
	at org.apache.flink.runtime.scheduler.declarative.WaitingForResources.createExecutionGraphWithAvailableResources(WaitingForResources.java:109) ~[classes/:?]
	at org.apache.flink.runtime.scheduler.declarative.WaitingForResources.resourceTimeout(WaitingForResources.java:103) ~[classes/:?]
	at org.apache.flink.runtime.scheduler.declarative.DeclarativeScheduler.runIfState(DeclarativeScheduler.java:870) ~[classes/:?]
	at org.apache.flink.runtime.scheduler.declarative.DeclarativeScheduler.lambda$runIfState$20(DeclarativeScheduler.java:885) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
16492 [pool-4-thread-1] ERROR org.apache.flink.runtime.util.FatalExitExceptionHandler [] - FATAL: Thread 'pool-4-thread-1' produced an uncaught exception. Stopping the process...
java.lang.AssertionError: null
	at org.apache.flink.runtime.concurrent.ComponentMainThreadExecutorServiceAdapter.lambda$new$0(ComponentMainThreadExecutorServiceAdapter.java:54) ~[test-classes/:?]
	at org.apache.flink.runtime.concurrent.ComponentMainThreadExecutorServiceAdapter.assertRunningInMainThread(ComponentMainThreadExecutorServiceAdapter.java:84) ~[test-classes/:?]
	at org.apache.flink.runtime.executiongraph.ExecutionGraph.assertRunningInJobMasterMainThread(ExecutionGraph.java:1469) ~[classes/:?]
	at org.apache.flink.runtime.executiongraph.ExecutionGraph.transitionState(ExecutionGraph.java:1052) ~[classes/:?]
	at org.apache.flink.runtime.executiongraph.ExecutionGraph.transitionState(ExecutionGraph.java:1044) ~[classes/:?]
	at org.apache.flink.runtime.executiongraph.ExecutionGraph.transitionToRunning(ExecutionGraph.java:834) ~[classes/:?]
	at org.apache.flink.runtime.scheduler.declarative.DeclarativeScheduler.createExecutionGraphWithAvailableResources(DeclarativeScheduler.java:580) ~[classes/:?]
	at org.apache.flink.runtime.scheduler.declarative.WaitingForResources.createExecutionGraphWithAvailableResources(WaitingForResources.java:109) ~[classes/:?]
	at org.apache.flink.runtime.scheduler.declarative.WaitingForResources.resourceTimeout(WaitingForResources.java:103) ~[classes/:?]
	at org.apache.flink.runtime.scheduler.declarative.DeclarativeScheduler.runIfState(DeclarativeScheduler.java:870) ~[classes/:?]
	at org.apache.flink.runtime.scheduler.declarative.DeclarativeScheduler.lambda$runIfState$20(DeclarativeScheduler.java:885) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
```

**Comment by zentol** [comment]
Oh god, the failure happens because of the DeclarativeSchedulerTest. That test fails the main thread validation checks and then terminates the JVM, killing other tests in the process...

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/SchedulerNGFactoryFactoryTest.java [hunk 6]
```diff
@@ -54,16 +57,29 @@ public void createSchedulerNGFactoryIfConfigured() {
...
         } catch (IllegalArgumentException e) {
-            assertThat(e.getMessage(), containsString("Illegal value [invalid-scheduler-name]"));
+            assertThat(
+                    e.getMessage(),
+                    containsString("Could not parse value 'invalid-scheduler-name'"));
         }
     }
 
+    @Test
+    public void fallBackIfBatchAndDeclarative() {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/SchedulerNGFactoryFactoryTest.java:72]: fall back to what? It is a good idea to state in the test name what the intended behaviour is.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerBuilder.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+                jobManagerJobMetricGroup,
+                shuffleMaster,
+                partitionTracker,
+                restartBackoffTimeStrategy,
+                new DefaultExecutionDeploymentTracker(),
+                System.currentTimeMillis(),
+                mainThreadExecutor,
+                error ->
+                        FatalExitExceptionHandler.INSTANCE.uncaughtException(
+                                Thread.currentThread(), error),
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerBuilder.java:178]: I guess we would want to make the fatal error handler also configurable in order to test the `runIfState` method properly.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+
+    @Test
+    public void testRunIfStateWithStateMismatch() throws Exception {
+        final DeclarativeScheduler scheduler =
+                new DeclarativeSchedulerBuilder(getJobGraph(), mainThreadExecutor).build();
+
+        AtomicBoolean ran = new AtomicBoolean(false);
+        scheduler.runIfState(new DummyState(), () -> ran.set(true));
+        assertThat(ran.get(), is(false));
+    }
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java:106]: I think we are missing a case where we test that failures are forwarded to the fatal error handler.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+
+        final ResourceCounter resourceRequirement =
+                ResourceCounter.withResource(ResourceProfile.UNKNOWN, 1);
+
+        assertThat(scheduler.hasEnoughResources(resourceRequirement), is(false));
+
+        offerSlots(
+                declarativeSlotPool, createSlotOffersForResourceRequirements(resourceRequirement));
+
+        assertThat(scheduler.hasEnoughResources(resourceRequirement), is(true));
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java:135]: I would split this into two tests.

Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java:135]: I think we should also test that non matched resources decrease `UNKNOWN` resources.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+                createSlotOffersForResourceRequirements(
+                        ResourceCounter.withResource(ResourceProfile.UNKNOWN, numAvailableSlots)));
+
+        final ExecutionGraph executionGraph =
+                scheduler.createExecutionGraphWithAvailableResources();
+
+        assertThat(
+                executionGraph.getJobVertex(JOB_VERTEX.getID()).getParallelism(),
+                is(numAvailableSlots));
+    }
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java:170]: Maybe also test that we try to restore from a given savepoint.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+
+    @Test
+    public void testStartScheduling() throws Exception {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java:173]:
```suggestion
    public void testStartSchedulingTransitionsToWaitingForResources() throws Exception {
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+        scheduler.startScheduling();
+
+        assertThat(
+                declarativeSlotPool.getResourceRequirements(),
+                contains(ResourceRequirement.create(ResourceProfile.UNKNOWN, PARALLELISM)));
+    }
+
+    /** Tests that the listener for new slots is properly set up. */
+    @Test
+    public void testResourceAcquisitionTriggersExecution() throws Exception {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java:208]:
```suggestion
    public void testResourceAcquisitionTriggersJobExecution() throws Exception {
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/allocator/SharedSlotTest.java [hunk 13]
```diff
@@ -166,6 +168,48 @@ public void testReleaseForbidsSubsequentLogicalSlotAllocations() {
...
+                        }));
+        logicalSlot2.tryAssignPayload(
+                new TestLogicalSlotPayload(
+                        cause -> {
+                            if (logicalSlot1.isAlive()) {
+                                logicalSlot1.releaseSlot(cause);
+                            }
+                        }));
+
+        sharedSlot.release(new Exception("test"));
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/allocator/SharedSlotTest.java:196]: What are we asserting here?

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for creating this PR @zentol. The changes look really good. I think what we can add are a bit more tests for some of the `DeclarativeScheduler` methods:

* stopWithSavepoint: What happens if we are not in a `StateWithExecutionGraph`
* updateTaskExecutionState: What happens if we are not in a `StateWithExecutionGraph`
* requestNextInputSplit: What happens if we are not in a `StateWithExecutionGraph`
* requestPartitionState: What happens if we are not in a `StateWithExecutionGraph`
* deliverOperatorEventToCoordinator: What happens if we are not in a `StateWithExecutionGraph`
* deliverCoordinationRequestToCoordinator: What happens if we are not in a `StateWithExecutionGraph`
* canScaleUp: calling the `ScaleUpController` only if the new parallelism is higher than the current one
* onFinished: Stops the checkpoint services and notifies the job status listener
* transitionToState: Making sure that we only set a new state if it is really new and that we are calling the `onLeave` and `onEnter` methods

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SingleLogicalSlot.java [hunk 14]
```diff
@@ -197,20 +196,21 @@ private void signalPayloadRelease(Throwable cause) {
...
-                    markReleased();
-
-                    if (throwable != null) {
-                        releaseFuture.completeExceptionally(throwable);
-                    } else {
-                        releaseFuture.complete(null);
-                    }
-                });
+        FutureUtils.assertNoException(
+                terminalStateFuture.whenComplete(
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SingleLogicalSlot.java:200]: hmm...AFAIK this will never be completed exceptionally, but then we also shouldn't handle Throwables like we do here (and instead fail with an IllegalStateException I suppose).

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/JobGraphJobInformation.java [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+
+    public JobID getJobID() {
+        return jobID;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    // note: the SlotSharingGroup returned is mutable.
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/JobGraphJobInformation.java:63]: the comment must've slipped down during a rebase, I'll move it to the interface.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/allocator/SharedSlotTest.java (hunk 13 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/allocator/SharedSlotTest.java:197]: That the release did not fail with an exception; I'll add some assertios to make this more explicit

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java (hunk 9 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java:135]: > I think we should also test that non matched resources decrease UNKNOWN resources.

Isn't that slot pool behavior?

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPoolServiceFactory.java [hunk 16]
```diff
@@ -34,14 +35,19 @@
...
 
         if (ClusterOptions.isDeclarativeResourceManagementEnabled(configuration)) {
+            if (ClusterOptions.isDeclarativeSchedulerEnabled(configuration)
+                    && jobType == JobType.STREAMING) {
+                return new DeclarativeSlotPoolServiceFactory(
+                        SystemClock.getInstance(), slotIdleTimeout, rpcTimeout);
+            }
 
             return new DeclarativeSlotPoolBridgeServiceFactory(
                     SystemClock.getInstance(), rpcTimeout, slotIdleTimeout, batchSlotTimeout);
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPoolServiceFactory.java:53]: Yes it's a bit of a mess; I'd like to move that to a follow-up though.

[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPoolServiceFactory.java:53]: Because the approach is not obvious to me and I don't want to delay the skeleton further.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java [hunk 17]
```diff
@@ -0,0 +1,909 @@
...
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/** Declarative scheduler. */
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java:123]: I can't mock the CheckpointCoordinator since it is created within the ExecutionGraph, and I don't see another way to test it that doesn't involve creating a genuine savepoint :/

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java (hunk 10 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java:170]: I can't mock the CheckpointCoordinator since it is created within the ExecutionGraph, and I don't see another way to test it that doesn't involve creating a genuine savepoint :/

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPoolServiceFactory.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPoolServiceFactory.java:53]: Sounds good.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SingleLogicalSlot.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SingleLogicalSlot.java]: Yes, we should fail hard in this case.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java (hunk 10 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java:239]: True. Damn it. That's currently the price we have to pay for directly using the `EG` and `CC`. Then let's not do it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java [hunk 18]
```diff
@@ -291,12 +480,127 @@ public void testHowToHandleFailureUnrecoverableFailure() throws Exception {
...
+    @Test
+    public void testDeliverOperatorEventToCoordinatorFailsInIllegalState() throws Exception {
+        final DeclarativeScheduler scheduler =
+                new DeclarativeSchedulerBuilder(createJobGraph(), mainThreadExecutor).build();
+
+        try {
+            scheduler.deliverOperatorEventToCoordinator(
+                    new ExecutionAttemptID(), new OperatorID(), new TestOperatorEvent());
+            fail("Should have failed with an exception.");
+        } catch (TaskNotRunningException expected) {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java:516]: maybe `@Test(expected = TaskNotRunningException.class)`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java [hunk 19]
```diff
@@ -291,12 +480,127 @@ public void testHowToHandleFailureUnrecoverableFailure() throws Exception {
...
+
+    @Test
+    public void testRequestNextInputSplitFailsInIllegalState() throws Exception {
+        final DeclarativeScheduler scheduler =
+                new DeclarativeSchedulerBuilder(createJobGraph(), mainThreadExecutor).build();
+
+        try {
+            scheduler.requestNextInputSplit(JOB_VERTEX.getID(), new ExecutionAttemptID());
+            fail("Should have failed with an exception.");
+        } catch (IOException expected) {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java:555]: Maybe `@Test(expected = IOException.class)`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java [hunk 20]
```diff
@@ -291,12 +480,127 @@ public void testHowToHandleFailureUnrecoverableFailure() throws Exception {
...
+
+    @Test
+    public void testRequestPartitionStateFailsInIllegalState() throws Exception {
+        final DeclarativeScheduler scheduler =
+                new DeclarativeSchedulerBuilder(createJobGraph(), mainThreadExecutor).build();
+
+        try {
+            scheduler.requestPartitionState(new IntermediateDataSetID(), new ResultPartitionID());
+            fail("Should have failed with an exception.");
+        } catch (PartitionProducerDisposedException expected) {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java:567]: Maybe `@Test(expected = PartitionProducerDisposedException.class)`.

**Review by tillrohrmann — APPROVED** [review]
Thanks for addressing my comments @zentol. LGTM. +1 for merging.
